framework/auth/views.py

```diff
@@ -167,21 +167,30 @@ def confirm_email_get(**kwargs):
     user = User.load(kwargs['uid'])
     token = kwargs['token']
     if user:
-        if user.confirm_email(token):  # Confirm and register the user
-            user.date_last_login = datetime.datetime.utcnow()
-            user.save()
-
-            # Go to settings page
-            status.push_status_message(language.WELCOME_MESSAGE, 'success')
-            response = redirect('/settings/')
-
-            return framework.auth.authenticate(user, response=response)
-    # Return data for the error template
-    return {
-        'code': http.BAD_REQUEST,
-        'message_short': 'Link Expired',
-        'message_long': language.LINK_EXPIRED
-    }, http.BAD_REQUEST
+        if user.confirm_email(token):
+            if user.unconfirmed_username:
+                old_username = user.username
+                user.username = user.unconfirmed_username
+                user.save()
+                # Emit signal that username changed
+                # Must send previous username for mailchimp subscriber lookup
+                signals.username_changed.send(user, old_username=old_username)
+                status.push_status_message(language.UPDATED_EMAIL_CONFIRMATION, 'success')
+                response = redirect('/settings/')
```

Comment:
Use `web_url_for('user_profile')` in preference to hard-coding a URL.

ref: http://cosdev.readthedocs.org/en/latest/osf/guidelines.html?highlight=web_url_for